src/audio/copier/copier.c

```diff
@@ -998,9 +998,23 @@ static int copier_comp_trigger(struct comp_dev *dev, int cmd)
 		return PPL_STATUS_PATH_STOP;
 
 	for (i = 0; i < cd->endpoint_num; i++) {
-		ret = cd->endpoint[i]->drv->ops.trigger(cd->endpoint[i], cmd);
-		if (ret < 0)
-			break;
+		if (dev->ipc_config.type == SOF_COMP_HOST && !cd->ipc_gtw) {
+			ret = comp_set_state(cd->endpoint[i], cmd);
+			if (ret < 0)
+				break;
+
+			if (ret == COMP_STATUS_STATE_ALREADY_SET) {
+				ret = PPL_STATUS_PATH_STOP;
+				break;
+			}
+			ret = host_zephyr_trigger(cd->hd, dev, cmd);
```

Comment:
same here. I'd use a `bool host_trigger = false;` variable here, set it here to `true` and check it and trigger after the loop.

Reply:
this stage still need use host dev, move out looks not so good I can move out at final patch.

Reply:
for host free/reset/prepare, in the last patch, all moved out from loop for better handling.